[dagster-embedded-elt][dlt] Documentation for dlt integration (#20763)
dagster-io/dagster: docs/content/integrations/embedded-elt/dlt.mdx

---
title: "Dagster Embedded ELT: dlt"
description: Ingest data with ease using Dagster and dlt
---

## dlt

To get started with `dagster-embedded-elt` and _dlt_, first familiarize yourself with the quick [introduction to dlt](https://dlthub.com/docs/intro). When using the _dlt_ integration, you can either write your own sources, or use one of the many pre-made [verified sources](https://dlthub.com/docs/dlt-ecosystem/verified-sources/) in the _dlt_ ecosystem.

> ## 1. Configuring your Dagster project to support `dlt`

Reviewer: I think this jumps too quickly into 'tutorial' mode. For people who are not wanting to do a tutorial, we should do a better job explaining the high level concepts they should be familiar with.

Author: Thank you - I added a background section, but it may need more. We'll see!

```bash
$ cd dlt_sources/

$ dlt init github snowflake
Looking up the init scripts in https://github.com/dlt-hub/verified-sources.git...
Cloning and configuring a verified source github (Source that load github issues, pull requests and reactions for a specific repository via customizable graphql query. Loads events incrementally.)
Do you want to proceed? [Y/n]: y

Verified source github was added to your project!
* See the usage examples and code snippets to copy from github_pipeline.py
* Add credentials for snowflake and other secrets in ./.dlt/secrets.toml
* requirements.txt was created. Install it with:
pip3 install -r requirements.txt
* Read https://dlthub.com/docs/walkthroughs/create-a-pipeline for more information
```

This has pulled down the code required to easily collect data from the GitHub API. It has also created a `requirements.txt` and a `.dlt/` configuration. In this case, we can remove these files, as we will be handling configuration with environment variables, and the requirements are to be defined in Dagster's dependencies specification.

> ```bash

Reviewer: wonder if it's better to just do the `rm` command rather than show the files?

Author: I think it may be beneficial to keep it, as you can reference the requirements.txt and `.dlt/secrets.yaml` to understand what needs to be added elsewhere in the project.

$ tree -a
.
├── .dlt               # can be removed
│   ├── .sources
│   ├── config.toml
│   └── secrets.toml
├── .gitignore
├── github
│   ├── README.md
│   ├── __init__.py
│   ├── helpers.py
│   ├── queries.py
│   └── settings.py
├── github_pipeline.py
└── requirements.txt   # can be removed
```

## 3. Define environment variables required by `dlt`

This integration manages connections and secrets using environment variables as `dlt`. The `dlt` library is able to "magically" infer required environment variables used by its sources and resources. See [Secrets and Configs](https://dlthub.com/docs/general-usage/credentials/configuration) for more information.

> In this example, the `github_reactions` source requires a GitHub access token, and the _snowflake_ destination requires database connection details. This results in the required environment variables:

Reviewer: `This results in the following required environment variables:`

```bash
SOURCES__GITHUB__ACCESS_TOKEN=""
DESTINATION__SNOWFLAKE__CREDENTIALS__DATABASE=""
DESTINATION__SNOWFLAKE__CREDENTIALS__PASSWORD=""
DESTINATION__SNOWFLAKE__CREDENTIALS__USERNAME=""
DESTINATION__SNOWFLAKE__CREDENTIALS__HOST=""
DESTINATION__SNOWFLAKE__CREDENTIALS__WAREHOUSE=""
DESTINATION__SNOWFLAKE__CREDENTIALS__ROLE=""
```

> Ensure that these are defined in your `.env` file, and in the Dagster Cloud environment.

Reviewer: ```suggestion
Ensure that these are defined in your environment
```

## 4. Define a `DagsterDltResource`

> 

Reviewer: Why am I doing this?

> Create a `DagsterDltResource` resources:

Reviewer: ```suggestion
Create a `DagsterDltResource` resource:
```

```python
from dagster_embedded_elt.dlt.resource import DagsterDltResource

dlt_resource = DagsterDltResource()
```

## 5. Create a `dlt_assets` definition for GitHub

> Using the `dlt_assets` decorator, we can specify the `dlt` GitHub source as the `dlt_source` parameter, and our pipeline specifying that we would like the data loaded to Snowflake in the `dlt_pipeline`. In the body of the decorated function we call the `dlt.run`

Reviewer: ```suggestion
The `dlt_assets` decorator takes a `dlt_source` and `dlt_pipeline` parameter. In this example, we will use the `github_reactions` source and create a `dlt_pipeline` to ingest from Github to Snowflake. 

We then yield from `dlt.run` in the body of the function which will execute the pipeline when this asset is materialized.
```

- **To read the [dlt introduction](https://dlthub.com/docs/intro)**, if you've never worked with dlt before.
- **To install the following libraries**:

  ```bash
  pip install dagster dagster-embedded-elt
  ```

  Refer to the [Dagster installation](/getting-started/install) guide for more info.

---

## Step 1: Configure your Dagster project to support dlt

The first step is to define a location for the `dlt` code used for ingesting data. We recommend creating a `dlt_sources` directory at the root of your Dagster project, but this code can reside anywhere within your Python project.

Run the following to create the `dlt_sources` directory:

```bash
cd $DAGSTER_HOME && mkdir dlt_sources
```

---

## Step 2: Initialize dlt ingestion code

In the `dlt_sources` directory, you can write your own ingestion code following the [dlt tutorial](https://dlthub.com/docs/tutorial/load-data-from-an-api) or you can use a verified source.

In this example, we'll use the [GitHub source](https://dlthub.com/docs/dlt-ecosystem/verified-sources/github) provided by `dlt`.

1. First, install the dlt command line utility by running:

   ```bash
   pip install dlt

>    ```

Reviewer 2: I added a prereqs section to this, which is standard for our guides. I feel like it would be better to add it in with installing the other libraries - what do you think?

Author: Actually, once a user installs `dagster-embedded-elt` `dlt` will be installed automatically. I will remove this item.

Reviewer 2: Oh good call! I keep forgetting we do this with our integration libraries.

└── requirements.txt   # can be removed
```

---

## Step 3: Define dlt environment variables

This integration manages connections and secrets using environment variables as `dlt`. The `dlt` library can infer required environment variables used by its sources and resources. Refer to [dlt's Secrets and Configs](https://dlthub.com/docs/general-usage/credentials/configuration) documentation for more information.

In the example we've been using:

- The `github_reactions` source requires a GitHub access token
- The Snowflake destination requires database connection details

> This results in the following required environment variables:

Reviewer 2: Where should users be defining these? In an `.env`? Elsewhere? We should be specific here if we can.

DESTINATION__SNOWFLAKE__CREDENTIALS__ROLE=""
```

Ensure that these variables are defined in your environment.

---

## Step 4: Define a DagsterDltResource

> The <PyObject module="dagster_embedded_elt.dlt" object="DagsterDltResource" /> provides a wrapper of a dlt pipeline runner. This can be shared across dlt pipelines and is a required parameter to our dlt asset definition.

Reviewer 2: What file should users be doing this in?

Author: This one I wasn't sure of, because it would be in a hypothetical `resource.py` file, or wherever the user typically defines their resources. In that sense, this "tutorial" isn't so much a tutorial as we don't prescribe the file structure. This could be changed, though.

```python
from dagster_embedded_elt.dlt import DagsterDltResource

dlt_resource = DagsterDltResource()
```

---

## Step 5: Create a dlt_assets definition for GitHub

> The `dlt_assets` decorator takes a `dlt_source` and `dlt_pipeline` parameter. In this example, we used the `github_reactions` source and created a `dlt_pipeline` to ingest data from Github to Snowflake:

Reviewer 2: What file should users be doing this in?

Also, we should show the imports being used as well. Code snippets should be as copy-and-pasteable as possible.
